src/mlpack/core/optimizers/svrg/svrg_impl.hpp

```diff
@@ -0,0 +1,176 @@
+ *
+ * Implementation of stochastic variance reduced gradient (SVRG).
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_SVRG_SVRG_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_SVRG_SVRG_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "svrg.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+template<typename UpdatePolicyType, typename DecayPolicyType>
+SVRGType<UpdatePolicyType, DecayPolicyType>::SVRGType(
+    const double stepSize,
+    const size_t batchSize,
+    const size_t maxIterations,
+    const size_t innerIterations,
+    const double tolerance,
+    const bool shuffle,
+    const UpdatePolicyType& updatePolicy,
+    const DecayPolicyType& decayPolicy,
+    const bool resetPolicy) :
+    stepSize(stepSize),
+    batchSize(batchSize),
+    maxIterations(maxIterations),
+    innerIterations(innerIterations),
+    tolerance(tolerance),
+    shuffle(shuffle),
+    updatePolicy(updatePolicy),
+    decayPolicy(decayPolicy),
+    resetPolicy(resetPolicy)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename UpdatePolicyType, typename DecayPolicyType>
+template<typename DecomposableFunctionType>
+double SVRGType<UpdatePolicyType, DecayPolicyType>::Optimize(
+    DecomposableFunctionType& function,
+    arma::mat& iterate)
+{
+  // Find the number of functions to use.
+  const size_t numFunctions = function.NumFunctions();
+
+  // To keep track of where we are and how things are going.
+  double overallObjective = 0;
+  double lastObjective = DBL_MAX;
+
+  // Set epoch length to n / b if the user asked for.
+  if (innerIterations == 0)
+    innerIterations = numFunctions;
+
+  // Initialize the update policy.
+  if (resetPolicy)
+    updatePolicy.Initialize(iterate.n_rows, iterate.n_cols);
+
+  // Now iterate!
+  arma::mat gradient(iterate.n_rows, iterate.n_cols);
+  arma::mat gradient0(iterate.n_rows, iterate.n_cols);
+  arma::mat iterate0;
+
+  // Find the number of batches.
+  size_t numBatches = numFunctions / batchSize;
+  if (numFunctions % batchSize != 0)
+    ++numBatches; // Capture last few.
+
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+      std::numeric_limits<size_t>::max() : maxIterations;
+  for (size_t i = 0; i < actualMaxIterations; ++i)
+  {
+    // Calculate the objective function.
+    overallObjective = 0;
+    for (size_t f = 0; f < numFunctions; f += batchSize)
+    {
+      const size_t effectiveBatchSize = std::min(batchSize, numFunctions - f);
+      overallObjective += function.Evaluate(iterate, f, effectiveBatchSize);
+    }
+
+    if (std::isnan(overallObjective) || std::isinf(overallObjective))
+    {
+      Log::Warn << "Katyusha: converged to " << overallObjective
```

Comment:
Oops, I think this should say `SVRG` not `Katyusha` (there are a couple messages like this, in `SARAH` too).